src/Microsoft.AspNetCore.Authentication.JwtBearer/JwtBearerHandler.cs

```diff
@@ -195,11 +202,103 @@ internal class JwtBearerHandler : AuthenticationHandler<JwtBearerOptions>
+                {
+                    // Only add a comma after the first param, if any
+                    builder.Append(',');
+                }
+                builder.Append(" error=\"invalid_token\", error_description=\"");
+                builder.Append(errorDescription);
+                builder.Append('\"');
+
+                Response.Headers.Append(HeaderNames.WWWAuthenticate, builder.ToString());
+            }
 
             return false;
         }
 
+        private static string CreateErrorDescription(Exception authFailure)
+        {
+            if (authFailure == null)
+            {
+                return string.Empty;
+            }
+
+            IEnumerable<Exception> exceptions;
+            if (authFailure is AggregateException)
+            {
+                var agEx = authFailure as AggregateException;
+                exceptions = agEx.InnerExceptions;
+            }
+            else
+            {
+                exceptions = new[] { authFailure };
+            }
+
+            var errorDescription = new StringBuilder();
+
+            foreach (var ex in exceptions)
+            {
+                // Order sensitive, some of these exceptions derive from others
+                // and we want to display the most specific message possible.
+                if (ex is SecurityTokenInvalidAudienceException)
+                {
+                    Append("The audience is invalid", errorDescription);
+                }
+                else if (ex is SecurityTokenInvalidIssuerException)
+                {
+                    Append("The issuer is invalid", errorDescription);
+                }
+                else if (ex is SecurityTokenNoExpirationException)
+                {
+                    Append("The token has no expiration", errorDescription);
+                }
+                else if (ex is SecurityTokenInvalidLifetimeException)
+                {
+                    Append("The token lifetime is invalid", errorDescription);
+                }
+                else if (ex is SecurityTokenNotYetValidException)
+                {
+                    Append("The token is not valid yet", errorDescription);
+                }
+                else if (ex is SecurityTokenExpiredException)
+                {
+                    Append("The token is expired", errorDescription);
+                }
+                else if (ex is SecurityTokenSignatureKeyNotFoundException)
+                {
+                    Append("The signature key was not found", errorDescription);
+                }
+                else if (ex is SecurityTokenInvalidSignatureException)
+                {
+                    Append("The signature is invalid", errorDescription);
+                }
+            }
+
+            return errorDescription.ToString();
+        }
+
+        private static void Append(string message, StringBuilder errorDescription)
```

Comment:
Nit: could consider just appending to a `List<string> errors` and then call String.Join(errors, ';') at the end, might be a little bit more compact without the Append helper.